src/units.cpp

```diff
@@ -126,43 +140,118 @@ std::vector<Unit>::iterator Units::UnitsImpl::findUnit(const std::string &refere
+        double exp;
+        double mult;
+        double expMult;
+        for (size_t i = 0; i < units->unitCount(); ++i) {
+            units->unitAttributes(i, ref, pre, exp, expMult, id);
+            mult = std::log10(expMult);
+            if (!isStandardUnitName(ref)) {
+                auto model = owningModel(units);
+                auto refUnits = model->units(ref);
+                updated = updateUnitMultipliers(unitMultiplierMap, multiplier, refUnits, exp * uExp, logMult + mult * uExp + standardPrefixList.at(pre) * uExp, direction);
+            } else {
+                for (const auto &iter : standardUnitsList.at(ref)) {
+                    unitMultiplierMap.at(iter.first) += direction * (iter.second * exp * uExp);
+                }
+                multiplier += direction * (logMult + (standardMultiplierList.at(ref) + mult + standardPrefixList.at(pre)) * exp);
+            }
+        }
+    } else if (unitMultiplierMap.find(unitsName) == unitMultiplierMap.end()) {
+        unitMultiplierMap.emplace(std::pair<std::string, double>(unitsName, direction * uExp));
+        multiplier += direction * logMult;
+        updated = true;
+    } else if (isStandardUnitName(unitsName)) {
+        for (const auto &iter : standardUnitsList.at(unitsName)) {
+            unitMultiplierMap.at(iter.first) += direction * (iter.second * uExp);
+        }
+        multiplier += direction * logMult;
+        multiplier += direction * standardMultiplierList.at(unitsName);
+        updated = true;
+    }
+
+    return updated;
 }
+*/
 
-Units::~Units()
+bool updateUnitMultipliers(double &multiplier,
+                           const UnitsPtr &units,
+                           double uExp, double logMult,
+                           int direction)
 {
-    delete mPimpl;
+    bool updated = false;
+    auto unitsName = units->name();
+    if (units->isBaseUnit()) {
+        multiplier += direction * logMult;
+        updated = true;
+    } else {
+        std::string ref;
+        std::string pre;
+        std::string id;
+        double exp;
+        double mult;
+        double expMult;
+        for (size_t i = 0; i < units->unitCount(); ++i) {
+            units->unitAttributes(i, ref, pre, exp, expMult, id);
+            mult = std::log10(expMult);
+            if (isStandardUnitName(ref)) {
+                multiplier += direction * (logMult + (standardMultiplierList.at(ref) + mult + standardPrefixList.at(pre)) * exp);
+            } else {
+                auto model = owningModel(units);
+                auto refUnits = model->units(ref);
+                updated = updateUnitMultipliers(multiplier, refUnits, exp * uExp, logMult + mult * uExp + standardPrefixList.at(pre) * uExp, direction);
+            }
+        }
+    }
+
+    return updated;
 }
 
-Units::Units(const Units &rhs)
-    : NamedEntity(rhs)
-    , ImportedEntity(rhs)
-    , mPimpl(new UnitsImpl())
+double doGetScalingFactor(const UnitsPtr &units1, const UnitsPtr &units2)
```

Comment:
This method is only used in `Units::scalingFactor()`, so I would the body of `doGetScalingFactor()` there.